realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -16,32 +16,60 @@
 
 package io.realm.internal;
 
-import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 
 public class Context {
 
+    // Pool to hold the phantom references.
+    // The size of array for storing phantom references will never decrease. Instead, we use another array to hold the
+    // index of the free slot. When adding the reference, pick the last index from freeIndexList and put the reference
+    // to the corresponding slot.
+    private static class ReferencesPool {
+        ArrayList<NativeObjectReference> arrayList = new ArrayList<>();
+        ArrayList<Integer> freeIndexList = new ArrayList<Integer>();
+
+        void add(NativeObjectReference ref) {
+            if (arrayList.size() <= ref.refIndex) {
+                arrayList.add(ref);
+            } else {
+                arrayList.set(ref.refIndex, ref);
+            }
+        }
+
+        Integer getFreeIndex() {
+            Integer index;
+            if (freeIndexList.isEmpty()) {
+                index = arrayList.size();
+            } else {
+                int freeIndexToRemove = freeIndexList.size() - 1;
+                index = freeIndexList.get(freeIndexToRemove);
+                freeIndexList.remove(freeIndexToRemove);
+            }
+            return index;
+        }
+    }
+
     // Each group of related Realm objects will have a Context object in the root.
     // The root can be a table, a group, or a shared group.
     // The Context object is used to store a list of native pointers 
     // whose disposal need to be handed over from the garbage 
     // collection thread to the users thread.
 
-    // Reserved to be used only as a placeholder by native references Map to avoid autoboxing allocations
-    static final Integer NATIVE_REFERENCES_VALUE = 0;
-
     private List<Long> abandonedTables = new ArrayList<Long>();
     private List<Long> abandonedTableViews = new ArrayList<Long>();
     private List<Long> abandonedQueries = new ArrayList<Long>();
 
-    HashMap<Reference<?>, Integer> rowReferences = new HashMap<Reference<?>, Integer>();
-    ReferenceQueue<NativeObject> referenceQueue = new ReferenceQueue<NativeObject>();
+    private ReferencesPool referencesPool = new ReferencesPool();
+    private ReferenceQueue<NativeObject> referenceQueue = new ReferenceQueue<NativeObject>();
 
     private boolean isFinalized = false;
 
+    public synchronized void addReference(int type, NativeObject referent) {
+        referencesPool.add(new NativeObjectReference(type, referent, referenceQueue, referencesPool.getFreeIndex()));
+    }
+
     public void executeDelayedDisposal() {
         synchronized (this) {
```

Comment:
I am not as versed as I should be about `executeDelayedDisposal()` but it involves with checking the sizes of three unprotected `ArrayList`; `abandonedTables`, `abandonedTableViews`, and `abandonedQueries`. 

Those three are handled without protection in `asyncDisposeTable()`, `asyncDisposeTableView()`, `asyncDisposeQuery()`. Although it is clear that they are not your former work, I would highly appreciate if you could  look into the methods and apply appropriate measures to protect the lists.